Common/src/linear_algebra/CSysMatrix.cpp

```diff
@@ -27,162 +27,160 @@
 #endif
 
 }
 
 template<class ScalarType>
-void CSysMatrix<ScalarType>::Initialize(unsigned long nPoint, unsigned long nPointDomain,
-                            unsigned short nVar, unsigned short nEqn,
+void CSysMatrix<ScalarType>::Initialize(unsigned long npoint, unsigned long npointdomain,
+                            unsigned short nvar, unsigned short neqn,
                             bool EdgeConnect, CGeometry *geometry, CConfig *config) {
 
-  /*--- Don't delete *row_ptr, *col_ind because they are
-   asigned to the Jacobian structure. ---*/
+  if(matrix != nullptr)
+    SU2_MPI::Error("CSysMatrix can only be initialized once.", CURRENT_FUNCTION);
 
-  unsigned long iPoint, *row_ptr, *col_ind, index, nnz, Elem, iVar;
-  unsigned short iNeigh, iElem, iNode, *nNeigh, *nNeigh_ilu;
-  vector<unsigned long>::iterator it;
-  vector<unsigned long> vneighs, vneighs_ilu;
+  if(nvar > MAXNVAR)
+    SU2_MPI::Error("nVar larger than expected, increase MAXNVAR.", CURRENT_FUNCTION);
 
-  /*--- Set the ILU fill in level --*/
+  /*--- Application of this matrix, FVM or FEM. ---*/
+  auto type = EdgeConnect? ConnectivityType::FiniteVolume : ConnectivityType::FiniteElement;
 
-  ilu_fill_in = config->GetLinear_Solver_ILU_n();
+  /*--- Types of preconditioner the matrix will be asked to build. ---*/
+  unsigned short sol_prec = config->GetKind_Linear_Solver_Prec();
+  unsigned short def_prec = config->GetKind_Deform_Linear_Solver_Prec();
+  unsigned short adj_prec = config->GetKind_DiscAdj_Linear_Prec();
+  bool adjoint = config->GetDiscrete_Adjoint();
 
-  /*--- Compute the number of neighbors ---*/
+  bool ilu_needed = (sol_prec==ILU) || (def_prec==ILU) || (adjoint && (adj_prec==ILU));
 
-  nNeigh = new unsigned short [nPoint];
-  for (iPoint = 0; iPoint < nPoint; iPoint++) {
+  /*--- Basic dimensions. ---*/
+  nVar = nvar;
+  nEqn = neqn;
+  nPoint = npoint;
+  nPointDomain = npointdomain;
 
-    if (EdgeConnect) {
-      nNeigh[iPoint] = (geometry->node[iPoint]->GetnPoint()+1);  // +1 -> to include diagonal element
-    }
-    else {
-      vneighs.clear();
-      for (iElem = 0; iElem < geometry->node[iPoint]->GetnElem(); iElem++) {
-        Elem =  geometry->node[iPoint]->GetElem(iElem);
-        for (iNode = 0; iNode < geometry->elem[Elem]->GetnNodes(); iNode++)
-          vneighs.push_back(geometry->elem[Elem]->GetNode(iNode));
-      }
-      vneighs.push_back(iPoint);
+  /*--- Get sparse structure pointers from geometry,
+   *    the data is managed by CGeometry to allow re-use. ---*/
 
-      sort(vneighs.begin(), vneighs.end());
-      it = unique(vneighs.begin(), vneighs.end());
-      vneighs.resize(it - vneighs.begin());
-      nNeigh[iPoint] = vneighs.size();
-    }
+  const auto& csr = geometry->GetSparsePattern(type,0);
 
-  }
+  row_ptr = csr.outerPtr();
+  col_ind = csr.innerIdx();
+  dia_ptr = csr.diagPtr();
+  nnz = csr.getNumNonZeros();
```

Comment:
This is the pattern coming from the CGeometry associated with the matrix, CGeometry does lazy construction of those patterns so there is no extra work or complicate logic on the config settings to determine types of solver etc.